### PR TITLE
[GridBuilder] Grids as service

### PR DESCRIPTION
**Pull request by loic425**
```php
<?php

declare(strict_types=1);

namespace App\Grid;

use App\Entity\Author\Author;
use Sylius\Bundle\GridBundle\AbstractGrid;
use Sylius\Bundle\GridBundle\Builder\Field\StringField;
use Sylius\Bundle\GridBundle\Builder\Filter;
use Sylius\Bundle\GridBundle\Builder\GridBuilderInterface;

class BackendAuthorGrid extends AbstractGrid
{
    public static function getName(): string
    {
        return 'app_backend_author';
    }

    public static function getResourceClass(): string
    {
        return Author::class;
    }

    public function buildGrid(GridBuilder $gridBuilder): void
    {
        $gridBuilder
            ->addField(StringField::create('firstName')
                ->setLabel('sylius.ui.first_name')
                ->setSortable(true)
            )
            ->addField(StringField::create('lastName')
                ->setLabel('sylius.ui.last_name')
                ->setSortable(true)
            )
            ->orderBy('lastName')
            ->addFilter(Filter::create('search', 'string')
                ->setOptions([
                    'fields' => ['firstName', 'lastName']
                ])
            )
            ->addActionGroup(MainActionGroup::create(CreateAction::create()))
            ->addActionGroup(ItemActionGroup::create(
                Action::create('showBooks', 'default')
                    ->setLabel('app.ui.show_books')
                    ->setIcon('book')
                    ->setOptions([
                        'link' => [
                            'route' => 'app_backend_book_index',
                            'parameters' => [
                                'criteria' => ['author' => 'resource.id'],
                            ],
                        ],
                    ])
                 ),
                 UpdateAction::create(),
                 DeleteAction::create(),
            ) 
        ;
    }
}

```

```php
<?php

declare(strict_types=1);

namespace App\Grid;

use App\Entity\Book;
use App\Entity\Nationality;
use Sylius\Bundle\GridBundle\AbstractGrid;
use Sylius\Bundle\GridBundle\Builder\Field;
use Sylius\Bundle\GridBundle\Builder\Filter;
use Sylius\Bundle\GridBundle\Builder\GridBuilderInterface;

class BookByAmericanAuthorsGrid extends AbstractGrid
{
    public static function getName(): string
    {
        return 'app_book_by_american_authors';
    }

    public static function getResourceClass(): string
    {
        return Book::class;
    }

    public function buildGrid(GridBuilderInterface $gridBuilder): void
    {
        $gridBuilder
            ->setRepositoryMethod('createAmericanBooksQueryBuilder')
            ->addFilter(Filter::create('title', 'string'))
            ->addFilter(Filter::create('author', 'entity')
                ->setFormOptions([
                    'class' => Nationality::class,
                ])
            )
            ->addFilter(Filter::create('nationality', 'entity')
                ->setOptions([
                    'fields' => ['author.nationality'],
                ])
                ->setFormOptions([
                    'class' => Nationality::class,
                ])
            )
            ->orderBy('title', 'asc')
            ->addField(Field::create('title', 'string')
                ->setLabel('Title')
                ->setSortable(true)
            )
            ->addField(Field::create('author', 'string')
                ->setLabel('Author')
                ->setPath('author.name')
                ->setSortable(true, 'author.name')
            )
            ->addField(Field::create('nationality', 'string')
                ->setLabel('Nationality')
                ->setPath('na.name')
                ->setSortable(true, 'na.name')
            )
            ->setLimits([10, 5, 15])
        ;
    }
}
```

### CODE REVIEW

**Comment by loic425** [comment]
@pamil I added everything left in my head for this feature.
I do not allow to extends an existing yaml/xml/php configuration.
So if a user wants to customize an existing Sylius admin/shop grids, either he uses the existing event system or he can completely redefines it with a new grid service.
The chain provider that I added in this PR gets the grid definition from the first provider which contains this definition.
I added a negative default priority to these grids providers to set the ServiceGridProvider at first position. That way, a user can add its custom one and this new provider will be at the first position in the chain provider.

File: src/Component/AbstractGrid.php [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    public function getDefinition(): Grid
+    {
+        $gridBuilder = GridBuilder::create(static::getName(), static::getResourceClass());
+
+        $this->buildGrid($gridBuilder);
+
+        return $gridBuilder->getDefinition();
+    }
+
+    protected function buildGrid(GridBuilderInterface $gridBuilder): void
```
Comment (mamazu) [inline src/Component/AbstractGrid.php:22]: Why isn't this function abstract. A grid without any definition doesn't make sense to me.

File: src/Component/GridRegistry.php [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace Sylius\Component\Grid;
+
+final class GridRegistry
+{
+    private array $grids;
+
+    public function __construct(\Traversable $grids)
+    {
+        $this->grids = iterator_to_array($grids);
```
Comment (mamazu) [inline src/Component/GridRegistry.php:13]: What's wrong with storing the iterator here?

**Review by mamazu — COMMENTED** [review]
I really love the idea. :heart: PHP over configuration every time. Hope this gets merged. Then we only need a tool for refactoring it. :D

File: src/Component/AbstractGrid.php (hunk 1 above)
Comment (loic425) [reply src/Component/AbstractGrid.php:22]: Yes I think you're right. I was inspired by the Symfony\Component\Form\AbstractType and it does not have any abstract methods.
Maybe I could make it public and add it on the grid interface and remove id from AbstractGrid.
@pamil WDT ?

File: src/Component/GridRegistry.php (hunk 2 above)
Comment (loic425) [reply src/Component/GridRegistry.php:13]: To make that possible
```php
return $this->grids[$code] ?? null;
```
I just followed the [Symfony documentation](https://symfony.com/doc/current/service_container/tags.html#tagged-services-with-index)

Comment (mamazu) [reply src/Component/GridRegistry.php:13]: Ah yeah that makes sense.

File: src/Component/AbstractGrid.php (hunk 1 above)
Comment (loic425) [reply src/Component/AbstractGrid.php]: @mamazu I added the buildGrid method on the grid interface 🎉

Comment (mamazu) [reply src/Component/AbstractGrid.php]: Great, then this should be resolved.

**Comment by loic425** [comment]
@pamil To celebrate the end of the week, could you do a first review of what you think about architecture overall?

File: src/Component/GridInterface.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+namespace Sylius\Component\Grid;
+
+use Sylius\Component\Grid\Builder\GridBuilderInterface;
+use Sylius\Component\Grid\Definition\Grid;
+
+interface GridInterface
+{
+    public static function getName(): string;
+
+    public static function getResourceClass(): string;
```
Comment (smoench) [inline src/Component/GridInterface.php:23]: A resource class is related to the SyliusResourceBundle, right? I think this should be moved to the ResourceBundle. There are also some other resource related options which has been implemented in the SyliusResourceBundle https://github.com/Sylius/SyliusResourceBundle/tree/master/src/Bundle/Grid

File: src/Component/GridRegistry.php (hunk 2 above)
Comment (smoench) [reply src/Component/GridRegistry.php:22]: Well, as this class is part of the component namespace I think it shouldn't rely on Symfony's `tagged_iterator` behavior. Instead it should only accept an array of `GridInterface` and manually index it by `getName`. The Bundle needs to implement a compiler pass.

File: src/Component/GridInterface.php (hunk 3 above)
Comment (loic425) [reply src/Component/GridInterface.php:23]: ResourceInterface is related to Sylius Resource component 
https://github.com/Sylius/SyliusResourceBundle/blob/master/src/Component/Model/ResourceInterface.php

File: src/Component/GridRegistry.php (hunk 2 above)
Comment (loic425) [reply src/Component/GridRegistry.php]: @smoench I think you're right. I improved the GridRegistry to be detached tagged_iterator (with a map) behaviour.

File: src/Component/GridRegistry.php [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    private array $grids = [];
+
+    public function __construct(\Traversable $grids)
+    {
+        foreach ($grids as $grid) {
+            $this->addGrid($grid);
+        }
+    }
+
+    public function addGrid(GridInterface $grid): void
```
Comment (loic425) [inline src/Component/GridRegistry.php:27]: @pamil Maybe just add & get instead of addGrid and getGrid...

**Comment by loic425** [comment]
@pamil @lchrusciel I reworked this PR to use the config builders in [this PR](https://github.com/Sylius/SyliusGridBundle/pull/155) So this one should be reviewed after. This one does not need Symfony 5.3 to be used, but it works with the same config builders.

File: src/Bundle/test/src/Grid/AuthorGrid.php [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+namespace App\Grid;
+
+use App\Entity\Author;
+use Sylius\Bundle\GridBundle\Grid\AbstractGrid;
+use Sylius\Bundle\GridBundle\Builder\Field\StringField;
+use Sylius\Bundle\GridBundle\Builder\Filter\Filter;
+use Sylius\Bundle\GridBundle\Builder\GridBuilderInterface;
+use Sylius\Bundle\GridBundle\Grid\ResourceAwareGridInterface;
+
+class AuthorGrid extends AbstractGrid implements ResourceAwareGridInterface
```
Comment (lchrusciel) [inline src/Bundle/test/src/Grid/AuthorGrid.php:23]:
```suggestion
final class AuthorGrid extends AbstractGrid implements ResourceAwareGridInterface
```

File: src/Bundle/test/src/Grid/BookByAmericanAuthorsGrid.php [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+use App\Entity\Book;
+use App\Entity\Nationality;
+use Sylius\Bundle\GridBundle\Builder\Field\StringField;
+use Sylius\Bundle\GridBundle\Builder\Filter\Filter;
+use Sylius\Bundle\GridBundle\Builder\GridBuilderInterface;
+use Sylius\Bundle\GridBundle\Grid\AbstractGrid;
+use Sylius\Bundle\GridBundle\Grid\ResourceAwareGridInterface;
+
+class BookByAmericanAuthorsGrid extends AbstractGrid implements ResourceAwareGridInterface
```
Comment (lchrusciel) [inline src/Bundle/test/src/Grid/BookByAmericanAuthorsGrid.php:24]:
```suggestion
final class BookByAmericanAuthorsGrid extends AbstractGrid implements ResourceAwareGridInterface
```

File: src/Bundle/test/src/Grid/BookByAmericanAuthorsGrid.php [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+                ->setFormOptions([
+                    'class' => Nationality::class,
+                ])
+            )
+            ->addFilter(Filter::create('nationality', 'entity')
+                ->setOptions([
+                    'fields' => ['author.nationality'],
+                ])
+                ->setFormOptions([
+                    'class' => Nationality::class,
```
Comment (lchrusciel) [inline src/Bundle/test/src/Grid/BookByAmericanAuthorsGrid.php:51]: We should extend this functionality to provide possibility to put parameter here

File: src/Bundle/test/src/Grid/BookByEnglishAuthorsGrid.php [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+use App\Entity\Book;
+use App\Entity\Nationality;
+use App\QueryBuilder\EnglishBooksQueryBuilder;
+use Sylius\Bundle\GridBundle\Builder\Field\StringField;
+use Sylius\Bundle\GridBundle\Builder\Filter\Filter;
+use Sylius\Bundle\GridBundle\Builder\GridBuilderInterface;
+use Sylius\Bundle\GridBundle\Grid\AbstractGrid;
+use Sylius\Bundle\GridBundle\Grid\ResourceAwareGridInterface;
+
+class BookByEnglishAuthorsGrid extends AbstractGrid implements ResourceAwareGridInterface
```
Comment (lchrusciel) [inline src/Bundle/test/src/Grid/BookByEnglishAuthorsGrid.php:25]:
```suggestion
final class BookByEnglishAuthorsGrid extends AbstractGrid implements ResourceAwareGridInterface
```

File: src/Bundle/test/src/Grid/BookGrid.php [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+use App\Entity\Author;
+use App\Entity\Book;
+use App\Entity\Nationality;
+use Sylius\Bundle\GridBundle\Builder\Field\StringField;
+use Sylius\Bundle\GridBundle\Builder\Filter\Filter;
+use Sylius\Bundle\GridBundle\Builder\GridBuilderInterface;
+use Sylius\Bundle\GridBundle\Grid\AbstractGrid;
+use Sylius\Bundle\GridBundle\Grid\ResourceAwareGridInterface;
+
+class BookGrid extends AbstractGrid implements ResourceAwareGridInterface
```
Comment (lchrusciel) [inline src/Bundle/test/src/Grid/BookGrid.php:25]:
```suggestion
final class BookGrid extends AbstractGrid implements ResourceAwareGridInterface
```

File: src/Component/Provider/ChainProvider.php [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+declare(strict_types=1);
+
+namespace Sylius\Component\Grid\Provider;
+
+use Sylius\Component\Grid\Definition\Grid;
+use Sylius\Component\Grid\Exception\UndefinedGridException;
+use Sylius\Component\Grid\Provider\GridProviderInterface;
+
+final class ChainProvider implements GridProviderInterface
```
Comment (lchrusciel) [inline src/Component/Provider/ChainProvider.php:20]: missing spec for this class

**Review by lchrusciel — COMMENTED** [review]
`src/Bundle/test/config/db.sql` should be removed I suppose

File: src/Bundle/Provider/ServiceGridProvider.php [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+namespace Sylius\Bundle\GridBundle\Provider;
+
+use Sylius\Bundle\GridBundle\Registry\GridRegistry;
+use Sylius\Component\Grid\Definition\ArrayToDefinitionConverterInterface;
+use Sylius\Component\Grid\Definition\Grid;
+use Sylius\Component\Grid\Exception\UndefinedGridException;
+use Sylius\Component\Grid\Provider\GridProviderInterface;
+
+final class ServiceGridProvider implements GridProviderInterface
```
[the remark this answers is not in the document]
Comment (loic425) [reply src/Bundle/Provider/ServiceGridProvider.php:22]: done

File: src/Component/Provider/ChainProvider.php (hunk 10 above)
Comment (loic425) [reply src/Component/Provider/ChainProvider.php:20]: done ✅

File: src/Bundle/Resources/config/services.xml [hunk 12]
```diff
@@ -28,10 +28,24 @@
...
             <argument>%sylius.grids_definitions%</argument>
+            <tag name="sylius.grid_provider" key="array" priority="-200"/>
+        </service>
+        <service id="sylius.grid.service_grid_provider" class="Sylius\Bundle\GridBundle\Provider\ServiceGridProvider">
+            <argument type="service" id="sylius.grid.array_to_definition_converter" />
+            <argument type="service" id="sylius.grid.grid_registry" />
+            <tag name="sylius.grid_provider" key="service" priority="-100"/>
+        </service>
+        <service id="sylius.grid.chain_provider" class="Sylius\Component\Grid\Provider\ChainProvider">
+            <argument type="tagged_iterator" tag="sylius.grid_provider" index-by="key" />
```
[the remark this answers is not in the document]
Comment (loic425) [reply src/Bundle/Resources/config/services.xml]: removed ✅

File: src/Bundle/Grid/AbstractGrid.php [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\GridBundle\Grid;
+
+use Sylius\Bundle\GridBundle\Builder\GridBuilder;
+use Sylius\Bundle\GridBundle\Builder\GridBuilderInterface;
+
+abstract class AbstractGrid implements GridInterface
```
[the remark this answers is not in the document]
Comment (loic425) [reply src/Bundle/Grid/AbstractGrid.php:19]: We already have these 2 tests:
* https://github.com/Sylius/SyliusGridBundle/blob/d7cbebc898b13327b136c5fbb3e3fd704864da20/src/Bundle/Tests/DependencyInjection/GridBuilderConfigurationTest.php#L633
* https://github.com/Sylius/SyliusGridBundle/blob/d7cbebc898b13327b136c5fbb3e3fd704864da20/src/Bundle/Tests/DependencyInjection/GridBuilderConfigurationTest.php#L661

These two grids extend the abstract grid.

File: src/Bundle/Resources/config/services.xml [hunk 14]
```diff
@@ -28,10 +28,24 @@
         <service id="sylius.grid.array_to_definition_converter" class="Sylius\Component\Grid\Definition\ArrayToDefinitionConverter">
             <argument type="service" id="event_dispatcher" />
         </service>
-        <service id="sylius.grid.provider" class="Sylius\Component\Grid\Provider\ArrayGridProvider">
+        <service id="sylius.grid.grid_registry" class="Sylius\Bundle\GridBundle\Registry\GridRegistry">
+            <argument type="tagged_iterator" tag="sylius.grid" />
```
[the remark this answers is not in the document]
Comment (loic425) [reply src/Bundle/Resources/config/services.xml]: done 🌵

File: src/Bundle/test/src/Grid/AuthorWithBooksWithUseOutputWalkersDisabled.php [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+final class AuthorWithBooksWithUseOutputWalkersDisabled extends AbstractGrid implements ResourceAwareGridInterface
+{
+    public static function getName(): string
+    {
+        return 'app_author_with_books_with_use_output_walkers_disabled';
+    }
+
+    public static function getResourceClass(): string
+    {
+        return Author::class;
```
Comment (lchrusciel) [inline src/Bundle/test/src/Grid/AuthorWithBooksWithUseOutputWalkersDisabled.php:31]: This is another place which will block us from usage of this feature in Sylius itself :( But perhaps we could introduce registry, which will use non static function and use class name as parameters through injection

**Review by lchrusciel — APPROVED** [review]

**Comment by lchrusciel** [comment]
Thanks Loic! 🏅

**Comment by lchrusciel** [comment]
@mamazu I assume you will be able to migrate yaml to php in the near future. The question is, if anyone will have space to write some rector converter or something similar :)

**Comment by mamazu** [comment]
Already working on it. If someone knows how to implement it as part of reactor, get in touch. 
I currently have a first prototype that handles some cases here: https://github.com/mamazu/grid-config-converter

**Pull request merged**